### PR TITLE
[Tags] Add Events for add/delete/update, addTagToElement, removeTagFromElement

### LINKED ISSUE #5368: Tag add and removal events
# Feature Request
Currently there are no events dispatched on adding a tag to an element or removing a tag from an element.
In order to handle these events, regardless they're happening to a document or a data object, we'll need to dispatch TagEvents which contains the necessary data for the event listeners to handle it.

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Resolves #5368

### CODE REVIEW

File: models/Element/Tag.php [hunk 1]
```diff
@@ -98,7 +99,15 @@ public static function getTagsForElement($cType, $cId)
      */
     public static function addTagToElement($cType, $cId, Tag $tag)
     {
+        $event = new TagEvent($tag, [
+            'elementType' => $cType,
+            'elementId' => $cType,
```
Comment (jdreesen) [inline models/Element/Tag.php:104]:
```suggestion
            'elementId' => $cId,
```

File: models/Element/Tag.php [hunk 2]
```diff
@@ -110,7 +119,15 @@ public static function addTagToElement($cType, $cId, Tag $tag)
      */
     public static function removeTagFromElement($cType, $cId, Tag $tag)
     {
+        $event = new TagEvent($tag, [
+            'elementType' => $cType,
+            'elementId' => $cType,
```
Comment (jdreesen) [inline models/Element/Tag.php:124]:
```suggestion
            'elementId' => $cId,
```

File: models/Element/Tag.php [hunk 3]
```diff
@@ -338,4 +369,30 @@ public function correctPath()
...
+    {
+        try {
+            \Pimcore::getEventDispatcher()->dispatch(TagEvents::PRE_DELETE, new TagEvent($this));
+
+            $this->getDao()->delete();
+
+            \Pimcore::getEventDispatcher()->dispatch(TagEvents::POST_DELETE, new TagEvent($this));
+        } catch (\Exception $e) {
+            throw $e;
+        }
```
Comment (jdreesen) [inline models/Element/Tag.php:388]: The try-catch is unneeded if it only re-throws the exception.

```suggestion
        \Pimcore::getEventDispatcher()->dispatch(TagEvents::PRE_DELETE, new TagEvent($this));

        $this->getDao()->delete();

        \Pimcore::getEventDispatcher()->dispatch(TagEvents::POST_DELETE, new TagEvent($this));
```

**Review by jdreesen — CHANGES_REQUESTED** [review]

File: models/Element/Tag.php (hunk 3 above)
Comment (dvesh3) [reply models/Element/Tag.php]: yes, valid point. thanks!

**Review by brusch — APPROVED** [review]

File: lib/Event/Model/TagEvent.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+{
+    use ArgumentsAwareTrait;
+
+    /**
+     * @var Tag
+     */
+    protected $tag;
+
+    /**
+     * DataObjectEvent constructor.
```
Comment (jdreesen) [inline lib/Event/Model/TagEvent.php:31]:
```suggestion
     * TagEvent constructor.
```

Comment (dvesh3) [reply lib/Event/Model/TagEvent.php]: thnx!

**Pull request merged**